### DISCUSSION TITLE
zoomToElements Function Throws Error When Operating on a Large Number of Elements

### DISCUSSION THREAD

**Discussion by wangjing1111** [iTwin/itwinjs-core#7437] [Q&A] [answered]
I have an issue where the zoomToElements function throws an error when dealing with a large number of elements, for example, more than 15,000 elements. If the number of elements is smaller, the function works normally. The error and the code are as follows:
![1](https://github.com/user-attachments/assets/0ad0c575-2bca-47f5-bf9f-fda3f6adeb09)
![2](https://github.com/user-attachments/assets/24c77296-f561-4b64-9d91-2d1fc752725a)

**Comment by pmconne** [comment]
Hi @wangjing1111. You are almost certainly hitting a size limit. Why are you trying to zoom to 15,000 elements?

**Reply by wangjing1111** [reply]
The model elements I publish can reach tens of thousands. Due to business requirements, I need to select and zoom in on a portion of these elements based on professional categories to display them to users. Is there a limit to the number of elements for zoomToElements? What methods are available to help me achieve my needs?

**Reply by pmconne** [reply, marked as answer]
`zoomToElements` uses [getPlacementProps](https://www.itwinjs.org/reference/core-frontend/imodelconnection/imodelconnection/imodelconnection.elements/getplacements/) to build an ECSql query string that includes a `WHERE` criterion containing all of the Ids you supplied. There's a very high but finite limit on the length of that string. You're exceeding that limit.

Instead of calling `zoomToElements` directly with a huge list of element Ids, you can make multiple calls to `getPlacementProps` with much smaller lists of element Ids, accumulating a list of `Placement`s to pass to `zoomToPlacements`. If you wanted to be more efficient, you could avoid allocating that huge list of `Placement`s and instead simply compute the union of all of the bounding boxes. Either way, take a look at the [implementation of zoomToElements](https://github.com/iTwin/itwinjs-core/blob/ca78632beb3182b1f66aa028410c4a326e2782c0/core/frontend/src/Viewport.ts#L2186) to guide you.

**Reply by wangjing1111** [reply]
Thank you very much for your answer
